emission/analysis/userinput/matcher.py

```diff
@@ -85,56 +85,107 @@ def create_confirmed_objects(user_id):
+                update_confirmed_and_composite(last_confirmed_place)
+
+            if confirmed_tl is not None and not confirmed_tl.is_empty():
+                ts.bulk_insert(list(confirmed_tl), esta.EntryType.ANALYSIS_TYPE)
+
+            last_processed_ts = timeline.last_place().data.enter_ts
 
         epq.mark_confirmed_object_creation_done(user_id, last_processed_ts)
     except:
         logging.exception("Error while creating confirmed objects, timestamp is unchanged")
         epq.mark_confirmed_object_creation_failed(user_id)
 
+def create_and_link_timeline(ts, timeline, last_confirmed_place):
+    keys = eac.get_config()["userinput.keylist"]
+    confirmed_places = []
+    curr_confirmed_start_place = last_confirmed_place
+    if last_confirmed_place is None:
+        # If it is not present - maybe this user is getting started for the first
+        # time, we create an entry based on the first trip from the timeline
+        curr_confirmed_start_place = create_confirmed_entry(ts,
+            timeline.first_place(), esda.CONFIRMED_PLACE_KEY, keys)
+        logging.debug("no last confirmed place found, created place with id %s" %
+            curr_confirmed_start_place.get_id())
+        confirmed_places.append(curr_confirmed_start_place)
+    else:
+        # we update it with the information from the matching cleaned place
+        matching_cleaned_place = timeline.first_place()
+        curr_confirmed_start_place["data"]["exit_ts"] = matching_cleaned_place.data.exit_ts
+        curr_confirmed_start_place["data"]["exit_fmt_time"] = matching_cleaned_place.data.exit_fmt_time
+        curr_confirmed_start_place["data"]["exit_local_dt"] = matching_cleaned_place.data.exit_local_dt
+        curr_confirmed_start_place["data"]["user_input"] = \
+            get_user_input_dict(ts, matching_cleaned_place, keys)
+        curr_confirmed_start_place["additions"] = \
+            esdt.get_additions_for_timeline_entry_object(ts, matching_cleaned_place)
+        logging.debug("Found existing last confirmed place, setting exit information to %s, and trimming additions to %s" %
+            (matching_cleaned_place.data.exit_fmt_time, len(curr_confirmed_start_place["data"]["additions"])))
+
+    if curr_confirmed_start_place is None:
+        logging.debug("did not find any cleaned place either, early return")
+        return None
+
+    confirmed_trips = []
+
+    logging.debug("Iterating through %s cleaned trips to create confirmed trips" % len(timeline.trips))
+    for ctrip in timeline.trips:
+        # create the trip-like object
+        logging.debug("trip-like entry = %s" % ctrip)
+        if ctrip['metadata']['key'] == esda.EXPECTED_TRIP_KEY:
+            curr_confirmed_trip = create_confirmed_entry(ts, ctrip, esda.CONFIRMED_TRIP_KEY, keys)
+        elif ctrip['metadata']['key'] == esda.CLEANED_UNTRACKED_KEY:
+            curr_confirmed_trip = create_confirmed_entry(ts, ctrip, esda.CONFIRMED_UNTRACKED_KEY, keys)
```

Comment:
This part got me a little stuck during reading. I believe I understand what is going on, but a comment explaining why we are checking explicitly for either ctrips that contain EXPECTED_TRIP_KEY or CLEANED_UNTRACKED_KEY could help for code readability
```python
# This is where we convert the trip entries in the list to confirmed trips, and if the entry is untracked then we processed the untracked trip without populating all the necessary  confirmed_object_data
```

Reply:
This depends on an understanding of the data model.
The trip-like inputs to this timeline are EXPECTED TRIP and CLEANED_UNTRACKED